### ISSUE TITLE
freezing BatchNorm according to different variable

### ISSUE THREAD

**Issue by fsk119** [tensorpack/tensorpack#1183]
Hi, I get stuck into freezing BatchNorm according to different variable.
My logic is here:
```
if is_training:
    fakeIm = self.decoder(realz) # realz from uniform distribution
    realIm = self.decoder(fakez) # fakez  = self.encoder(im)
    outIm = self.decoder(outz)   # outz from another distribution

    loss = self.classifier(realIm, outIm) #
```
In my implementation, there are BN layers in the decoder.  So when training, decoding outz will influence moving summary of bias and scale, which will destroy performance of decoder. 
Could you give me some advise? Is it possible to train Gan and classifier together?

Thanks for your examples about gan and fantastic framework! It really really helps me a lot.

**Comment by ppwwyyxx** [comment]
I get that you didn't want to update moving statistic in ` decoder(outz)`, but you didn't say what you want to do instead.
The batchnorm layer can be called with `training=False` and it won't update moving statistics and details in https://tensorpack.readthedocs.io/modules/models.html#tensorpack.models.BatchNorm
If this is not what you want to do, you can also implement your own variants of batchnorm instead.

**Comment by fsk119** [comment]
Thanks for your reply.
Emmm, I plan to update moving statistics according to `fakeIm = self.decoder(realz)` and `realIm = self.decoder(fakez)`.  According to your advise, I overwrite the logic, 
```
if is_training:
    fakeIm = self.decoder(realz, is_training) # realz from uniform distribution
    realIm = self.decoder(fakez, is_training) # fakez  = self.encoder(im)
    outIm = self.decoder(outz, False)   # outz from another distribution

    loss = self.classifier(realIm, outIm) #
```
It does work!

**Comment by ppwwyyxx** [comment]
You can also use [argscope](https://tensorpack.readthedocs.io/tutorial/symbolic.html#argscope-and-linearwrap) around the last `decoder()` to achieve the same.

**Comment by ppwwyyxx** [comment]
> I plan to update moving statistics according to fakeIm = self.decoder(realz) and realIm = self.decoder(fakez)

I get that part, but you still have two choices in `self.decoder(outz)`. Even when you don't update moving statistics there, you can:
1. use the moving statistics (updated by other decoder()) to normalize
2. use the batch input statistics (i.e., statistics from outz) to normalize

Setting `training=False` will give you option 1.
Option 2 requires the latest commit https://github.com/tensorpack/tensorpack/commit/f6ede6124ef5414ee4761e0e4725b8c0cfc1319c, and it corresponds to `training=True, ema_update="skip"`. I implement this just now and haven't tested enough on it. 
You can read more at the updated docs: https://tensorpack.readthedocs.io/modules/models.html#tensorpack.models.BatchNorm
